pysmt/cnf.py

```diff
@@ -0,0 +1,217 @@
+                cnf.append(clause)
+        return k, frozenset(cnf)
+
+    def walk_or(self, formula, args):
+        if len(args) == 1:
+            return args[0]
+        k = self._key_var(formula)
+        cnf = [frozenset([self.mgr.Not(k)] + [a for a,_ in args])]
+        for a,c in args:
+            cnf.append(frozenset([k, self.mgr.Not(a)]))
+            for clause in c:
+                cnf.append(clause)
+        return k, frozenset(cnf)
+
+    def walk_not(self, formula, args):
+        a, cnf = args[0]
+        if a.is_true():
+            return self.mgr.FALSE(), CNFizer.TRUE_CNF
+        elif a.is_false():
+            return self.mgr.TRUE(), CNFizer.TRUE_CNF
+        else:
+            k = self._key_var(formula)
+            return k, cnf | frozenset([frozenset([self.mgr.Not(k),
+                                                  self.mgr.Not(a).simplify()]),
+                                       frozenset([k, a])])
+
+    def walk_implies(self, formula,  args):
+        a, cnf_a = args[0]
+        b, cnf_b = args[1]
+
+        k = self._key_var(formula)
+        not_a = self.mgr.Not(a).simplify()
+        not_b = self.mgr.Not(b).simplify()
+
+        return k, (cnf_a | cnf_b | frozenset([frozenset([not_a, b, k]),
+                                              frozenset([a, k]),
+                                              frozenset([not_b, k])]))
+
+    def walk_iff(self, formula, args):
+        a, cnf_a = args[0]
+        b, cnf_b = args[1]
+
+        k = self._key_var(formula)
+        not_a = self.mgr.Not(a).simplify()
+        not_b = self.mgr.Not(b).simplify()
+        not_k = self.mgr.Not(k)
+
+        return k, (cnf_a | cnf_b | frozenset([frozenset([not_a, not_b, k]),
+                                              frozenset([not_a, b, not_k]),
+                                              frozenset([a, not_b, not_k]),
+                                              frozenset([a, b, k])]))
+
+    def walk_symbol(self, formula,  args):
+        if formula.is_symbol(BOOL):
+            return formula, CNFizer.TRUE_CNF
+        else:
+            return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_function(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_real_constant(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_bool_constant(self, formula,  args):
+        if formula.is_true():
+            return formula, CNFizer.TRUE_CNF
+        else:
+            return formula, CNFizer.TRUE_CNF
+
+    def walk_int_constant(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_plus(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_minus(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_times(self, formula,  args):
+        return CNFizer.THEORY_PLACEHOLDER
+
+    def walk_equals(self, formula, args):
+        assert all(a == CNFizer.THEORY_PLACEHOLDER for a in args)
```

Comment:
Sure. Equals, LT, LE are boolean atoms, therefore can appear as leaves in the CNF, while Plus & Co are terms that cannot appear in the CNF.